text/0019-stricter-validation.md

```diff
@@ -0,0 +1,77 @@
+
+In sum: To see why the second case is acceptable, we have to understand the interaction between permissive mode and strict mode. This is subtle and difficult to explain. Instead, we want to implement strict mode as a standalone feature, which does not depend on features only present in permissive mode.
+
+## Detailed design
+
+We propose to change strict-mode validation so that it has the following features taken from permissive mode:
+* singleton boolean `True`/`False` types (which have 1/ special consideration of them with `&&`, `||`, and conditionals, and 2/ subtyping between them and `Boolean`, as is done now for permissive mode).
+* depth subtyping, but not width subtyping, to support examples like `{ foo: True } <: { foo: bool }`.
+* a requirement that arguments to `==` have the same type _unless_ we know that comparison will always be false, in which case we can give the `==` type `False`. The same goes for `contains`, `containsAll`, and other constructs that use `==` under the covers.
+* a requirement that branches of a conditional have the same type _unless_ we can simplify the conditional due to the guard having a `True`/`False` type.
+* (no union types)
+
+The current way we implement strict mode will change to accommodate these additions. We propose the following:
+
+1. Change the Rust code to implement strict mode by reusing the logic of permissive mode​ and using a flag to perform more restrictive checks. For example, when given `User` and `Admin` entity types, the code performing the least-upper-bound computation will return union type `User|Admin` in permissive mode, and will fail in strict mode. And, the code for conditionals would force branches' types to match in strict mode, but not in permissive mode. Etc.
+2. When checking in strict mode, we construct a new AST that performs the needed transformations, as in step 2 today. 
+3. Either way, we can add type annotations. For strict mode, they could be added during the transformation; for permissive mode, no transformation is needed. No checking is done after annotations are added, since the checking is always done, regardless of mode, prior to the transformation.
+
+Permissive mode validation considers different `Action`s to have distinct types, e.g., `Action::"view"` and `Action::"update"` don't have the same type. This is in anticipation supporting distinct actions having distinct attribute sets. Today, the least upper bound of any two actions is the type _AnyEntity_, which is a general supertype of all entities, used internally. Such a type supports expressions that are found in the policy scope such as `action in [Action::"view", Action::"edit"]`. This expression typechecks because the least upper bound of `Action::"view"` and `Action::"edit"` is _AnyEntity_ so the expression is type `Set<`_AnyEntity_`>`. Therefore in our amended strict mode we either need to:
```

Comment:
The `action in [Action::"a", Action::"b"]` case discussed here does show up as a few failing test cases. Treating `AnyEntity` as part of "depth subtyping" seems wrong - it definitely applies width subtyping. I'm in favor of forcing all actions to share one entity type. We might even be able to make this work by allowing depth subtyping between the type of concrete attributes for a particular action and declared attribute types. This way we would still give `Action::"view".is_read_only` type `False` (if it is defined `false`) while being able to support a LUB between `Action::"view"` and `Action::"read"` where their LUB would be `Action` in which `is_read_only : Bool`.

Reply:
To be clear: You are saying that anticipated future support for attributes of `Action` entities would require (1) defining an `Action` type in the `entityTypes` part of the schema that tabulates the possible attributes and their types, and (2) having individual actions in the `actions` part define values of those attributes, which are checked against the `Action` type. Your comment about depth subtyping is that you might infer a more specific type for individual actions that is a subtype of the general type. Is this correct?

If so, I'm good with this, too.

Reply:
Yes. The goal is that boolean, and later string/long enum, typed action attributes can have a more specific type inferred while still requiring that all the same attributes are defined.

Reply:
decisions on this: They must have the same action type. For the purpose of Dafny, this is an action type without any attributes since it doesn't know about action attributes